README.md
# PakhshKit JS Google Analytics - Google Analytics plugin for the [Vidiun Player JS]

[![Build Status](https://travis-ci.org/vidiun/pakhshkit-js-google-analytics.svg?branch=master)](https://travis-ci.org/vidiun/pakhshkit-js-google-analytics)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![](https://img.shields.io/npm/v/@pakhshkit-js/pakhshkit-js-google-analytics/latest.svg)](https://www.npmjs.com/package/@pakhshkit-js/pakhshkit-js-google-analytics)
[![](https://img.shields.io/npm/v/@pakhshkit-js/pakhshkit-js-google-analytics/canary.svg)](https://www.npmjs.com/package/@pakhshkit-js/pakhshkit-js-google-analytics/v/canary)

PakhshKit JS Google Analytics plugin integrates google analytics with the [Vidiun Player JS].

PakhshKit JS Google Analytics is written in [ECMAScript6], statically analysed using [Flow] and transpiled in ECMAScript5 using [Babel].

[flow]: https://flow.org/
[ecmascript6]: https://github.com/ericdouglas/ES6-Learning#articles--tutorials
[babel]: https://babeljs.io

## Getting Started

### Prerequisites

The plugin requires [Vidiun Player JS] to be loaded first.

[vidiun player js]: https://github.com/vidiun/vidiun-player-js

### Installing

First, clone and run [yarn] to install dependencies:

[yarn]: https://yarnpkg.com/lang/en/

```
git clone https://github.com/vidiun/pakhshkit-js-google-analytics.git
cd pakhshkit-js-google-analytics
yarn install
```

### Building

Then, build the player

```javascript
yarn run build
```

### Embed the library in your test page

Finally, add the bundle as a script tag in your page, and initialize the player

```html
<script type="text/javascript" src="/PATH/TO/FILE/vidiun-{ovp/tv}-player.js"></script>
<script type="text/javascript" src="/PATH/TO/FILE/pakhshkit-google-analytics.js"></script>
<div id="player-placeholder" style="height:360px; width:640px">
<script type="text/javascript">
var config = {
  targetId: 'player-placeholder',
  provider: {
    partnerId: {PARTNER_ID}
    ...
  },
  player: {
   plugins: {
     googleAnalytics: {
       trackingId: 'UA-1234567-89'
     }
    ...
   }
   ...
  }
 ...
};
var player = VidiunPlayer.setup(config);
player.loadMedia({
  entryId: '{ENTRY_ID}'
  ...
});
</script>
```

## Documentation

- **[Configuration](docs/configuration.md)**

## Running the tests

Tests can be run locally via [Karma], which will run on Chrome, Firefox and Safari

[karma]: https://karma-runner.github.io/1.0/index.html

```
yarn run test
```

You can test individual browsers:

```
yarn run test:chrome
yarn run test:firefox
yarn run test:safari
```

### And coding style tests

We use ESLint [recommended set](http://eslint.org/docs/rules/) with some additions for enforcing [Flow] types and other rules.

See [ESLint config](.eslintrc.json) for full configuration.

We also use [.editorconfig](.editorconfig) to maintain consistent coding styles and settings, please make sure you comply with the styling.

## Compatibility

TBD

## Contributing

Please read [CONTRIBUTING.md](https://gist.github.com/PurpleBooth/b24679402957c63ec426) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/vidiun/pakhshkit-js-google-analytics/tags).

## License

This project is licensed under the AGPL-3.0 License - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.5"></a>
## [0.1.5](https://github.com/vidiun/pakhshkit-js-google-analytics/compare/v0.1.3...v0.1.5) (2019-01-30)


### Bug Fixes

* **FEC-8605:** change media event is triggered on first entry load ([#13](https://github.com/vidiun/pakhshkit-js-google-analytics/issues/13)) ([b1ab338](https://github.com/vidiun/pakhshkit-js-google-analytics/commit/b1ab338))



<a name="0.1.4"></a>
## [0.1.4](https://github.com/vidiun/pakhshkit-js-google-analytics/compare/v0.1.3...v0.1.4) (2019-01-30)


### Bug Fixes

* **FEC-8605:** change media event is triggered on first entry load ([#13](https://github.com/vidiun/pakhshkit-js-google-analytics/issues/13)) ([b1ab338](https://github.com/vidiun/pakhshkit-js-google-analytics/commit/b1ab338))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/vidiun/pakhshkit-js-google-analytics/compare/v0.1.2...v0.1.3) (2018-05-24)


### Bug Fixes

* **FEC-8253:** percentage Google analytics events are not sent in Firefox ([#8](https://github.com/vidiun/pakhshkit-js-google-analytics/issues/8)) ([ae5409a](https://github.com/vidiun/pakhshkit-js-google-analytics/commit/ae5409a))



<a name="0.1.2"></a>
## [0.1.2](https://github.com/vidiun/pakhshkit-js-google-analytics/compare/v0.1.1...v0.1.2) (2018-05-23)


### Bug Fixes

* **FEC-8230:** when playing live stream all "% watched" events are triggered ([#7](https://github.com/vidiun/pakhshkit-js-google-analytics/issues/7)) ([7a7df41](https://github.com/vidiun/pakhshkit-js-google-analytics/commit/7a7df41))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/vidiun/pakhshkit-js-google-analytics/compare/v0.1.0...v0.1.1) (2018-05-06)


### Bug Fixes

* consider functional category ([#6](https://github.com/vidiun/pakhshkit-js-google-analytics/issues/6)) ([ea7831b](https://github.com/vidiun/pakhshkit-js-google-analytics/commit/ea7831b))
* round the event value ([#4](https://github.com/vidiun/pakhshkit-js-google-analytics/issues/4)) ([1a27044](https://github.com/vidiun/pakhshkit-js-google-analytics/commit/1a27044))



<a name="0.1.0"></a>
# 0.1.0 (2018-04-25)


### Features

* **FEC-8036:** google analytics ([1b1d591](https://github.com/vidiun/pakhshkit-js-google-analytics/commit/1b1d591))


test/src/google-analytics.spec.js
import '../../src/index.js';
import {loadPlayer, FakeEvent, Error} from 'pakhshkit-js';
import * as TestUtils from 'pakhshkit-js/test/src/utils/test-utils';

const targetId = 'player-placeholder_google-analytics.spec';

describe('Google Analytics Plugin', function() {
  let player;
  const id = '1_rwbj3j0a';
  const partnerId = 1068292;
  const uiConfId = 123456;
  const entryName = 'some name';

  const CMid = '1_fdsfds78';
  const CMpartnerId = 876543;
  const CMuiConfId = 654321;
  const CMentryName = 'change media name';

  let config;

  const CMconfig = {
    CMid,
    sources: {
      progressive: [
        {
          mimetype: 'video/mp4',
          url: 'https://www.w3schools.com/tags/movie.mp4'
        }
      ]
    },
    plugins: {
      googleAnalytics: {
        trackingId: 'UA-1234567-89',
        entryId: CMid,
        entryName: CMentryName,
        uiConfId: CMuiConfId,
        partnerId: CMpartnerId
      }
    }
  };

  function createPlayerPlaceholder(targetId) {
    TestUtils.createElement('DIV', targetId);
    let el = document.getElementById(targetId);
    el.style.height = '360px';
    el.style.width = '640px';
  }

  function setupPlayer(config) {
    player = loadPlayer(config);
    const el = document.getElementById(targetId);
    el.appendChild(player.getView());
  }

  function verifyEventParams(eventData, eventParams = {}) {
    eventParams.category = eventParams.category || 'Vidiun Video Events';
    eventParams.label = eventParams.label || `${partnerId} | ${uiConfId} | ${id} | '${entryName}'`;
    eventData[2]['event_category'].should.equal(eventParams.category);
    if (eventData[2]['event_label']) {
      eventData[2]['event_label'].should.equal(eventParams.label);
    }
  }

  function verifyEventValue(eventData, value) {
    if (value) {
      eventData[2]['value'].should.equal(value);
    }
  }

  function verifyEventName(eventData, name) {
    eventData[1].should.equal(name);
  }

  before(function() {
    createPlayerPlaceholder(targetId);
  });

  afterEach(function() {
    player.destroy();
    player = null;
    TestUtils.removeVideoElementsFromTestPage();
  });

  after(function() {
    TestUtils.removeElement(targetId);
  });
  describe('default events', () => {
    beforeEach(function() {
      config = {
        id,
        sources: {
          progressive: [
            {
              mimetype: 'video/mp4',
              url: 'https://www.w3schools.com/tags/movie.mp4'
            }
          ]
        },
        plugins: {
          googleAnalytics: {
            trackingId: 'UA-1234567-89',
            entryId: id,
            entryName,
            uiConfId,
            partnerId
          }
        }
      };
      setupPlayer(config);
    });

    it('should send widget loaded', () => {
      verifyEventName(dataLayer[dataLayer.length - 1], 'widget loaded');
      verifyEventParams(dataLayer[dataLayer.length - 1]);
    });

    it('should send media ready', done => {
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'media ready');
        verifyEventParams(dataLayer[dataLayer.length - 1]);
        done();
      });
      player.load();
    });

    it('should send play & first play', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        verifyEventName(dataLayer[dataLayer.length - 2], 'play');
        verifyEventParams(dataLayer[dataLayer.length - 2]);
        verifyEventValue(dataLayer[dataLayer.length - 2], 1);
        verifyEventName(dataLayer[dataLayer.length - 1], 'first play');
        verifyEventParams(dataLayer[dataLayer.length - 1]);
        done();
      });
      player.play();
    });

    it('should send pause', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.addEventListener(player.Event.PAUSE, () => {
          verifyEventName(dataLayer[dataLayer.length - 1], 'pause');
          verifyEventParams(dataLayer[dataLayer.length - 1]);
          verifyEventValue(dataLayer[dataLayer.length - 1], 1);
          done();
        });
        player.pause();
      });
      player.play();
    });

    it('should send play', done => {
      player.addEventListener(player.Event.FIRST_PLAY, () => {
        player.addEventListener(player.Event.PAUSE, () => {
          player.addEventListener(player.Event.PLAY, () => {
            verifyEventName(dataLayer[dataLayer.length - 1], 'play');
            verifyEventParams(dataLayer[dataLayer.length - 1]);
            verifyEventValue(dataLayer[dataLayer.length - 1], 1);
            done();
          });
          player.play();
        });
        player.pause();
      });
      player.play();
    });

    it('should send seek', done => {
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        player.addEventListener(player.Event.SEEKED, () => {
          verifyEventName(dataLayer[dataLayer.length - 1], 'seek');
          verifyEventParams(dataLayer[dataLayer.length - 1]);
          verifyEventValue(dataLayer[dataLayer.length - 1], 2);
          done();
        });
        player.currentTime = 2;
      });
      player.load();
    });

    it('should send 25% watched', done => {
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        player.addEventListener(player.Event.SEEKED, () => {
          verifyEventName(dataLayer[dataLayer.length - 2], '25% watched');
          verifyEventParams(dataLayer[dataLayer.length - 2]);
          verifyEventValue(dataLayer[dataLayer.length - 1], 5);
          done();
        });
        player.currentTime = 5;
      });
      player.load();
    });

    it('should send 50% watched', done => {
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        player.addEventListener(player.Event.SEEKED, () => {
          verifyEventName(dataLayer[dataLayer.length - 2], '50% watched');
          verifyEventParams(dataLayer[dataLayer.length - 2]);
          verifyEventValue(dataLayer[dataLayer.length - 1], 7);
          done();
        });
        player.currentTime = 7;
      });
      player.load();
    });

    it('should send 75% watched', done => {
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        player.addEventListener(player.Event.SEEKED, () => {
          verifyEventName(dataLayer[dataLayer.length - 2], '75% watched');
          verifyEventParams(dataLayer[dataLayer.length - 2]);
          verifyEventValue(dataLayer[dataLayer.length - 1], 10);
          done();
        });
        player.currentTime = 10;
      });
      player.load();
    });

    it('should send 100% watched & ended', done => {
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        player.addEventListener(player.Event.SEEKED, () => {
          player.addEventListener(player.Event.ENDED, () => {
            verifyEventName(dataLayer[dataLayer.length - 3], '100% watched');
            verifyEventParams(dataLayer[dataLayer.length - 3]);
            verifyEventValue(dataLayer[dataLayer.length - 3], 13);
            verifyEventName(dataLayer[dataLayer.length - 1], 'ended');
            verifyEventParams(dataLayer[dataLayer.length - 1]);
            done();
          });
          player.play();
        });
        player.currentTime = 12;
      });
      player.load();
    });

    it('should send enter full screen', done => {
      player.addEventListener(player.Event.ENTER_FULLSCREEN, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'enter full screen');
        verifyEventParams(dataLayer[dataLayer.length - 1]);
        verifyEventValue(dataLayer[dataLayer.length - 1], 1);
        done();
      });
      player.notifyEnterFullscreen();
    });

    it('should send exit full screen', done => {
      player.addEventListener(player.Event.ENTER_FULLSCREEN, () => {
        player.addEventListener(player.Event.EXIT_FULLSCREEN, () => {
          verifyEventName(dataLayer[dataLayer.length - 1], 'exit full screen');
          verifyEventParams(dataLayer[dataLayer.length - 1]);
          verifyEventValue(dataLayer[dataLayer.length - 1], 1);
          done();
        });
        player.notifyExitFullscreen();
      });
      player.notifyEnterFullscreen();
    });

    it('should not send change media', done => {
      player.addEventListener(player.Event.CHANGE_SOURCE_ENDED, () => {
        done('sent changed media');
      });
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        player.addEventListener(player.Event.SEEKED, () => {
          player.addEventListener(player.Event.ENDED, () => {
            done();
          });
          player.play();
        });
        player.currentTime = 12;
      });
      player.load();
    });

    it('should send change media', done => {
      player.addEventListener(player.Event.CHANGE_SOURCE_ENDED, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'change media');
        verifyEventParams(dataLayer[dataLayer.length - 1], {label: `${CMpartnerId} | ${CMuiConfId} | ${CMid} | '${CMentryName}'`});
        verifyEventValue(dataLayer[dataLayer.length - 1], 1);
        done();
      });
      player.configure(CMconfig);
      player.configure(CMconfig);
    });

    it('should send critical error', done => {
      player.addEventListener(player.Event.ERROR, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'error');
        verifyEventParams(dataLayer[dataLayer.length - 1], {
          category: 'Vidiun Video Error',
          label: 'NO_SOURCE_PROVIDED'
        });
        done();
      });
      player.dispatchEvent(new FakeEvent('error', new Error(Error.Severity.CRITICAL, Error.Category.PLAYER, Error.Code.NO_SOURCE_PROVIDED)));
    });

    it('should not send non critical error', done => {
      player.addEventListener(player.Event.ERROR, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'widget loaded');
        done();
      });
      player.dispatchEvent(new FakeEvent('error', new Error(Error.Severity.RECOVERABLE, Error.Category.PLAYER, Error.Code.NO_SOURCE_PROVIDED)));
    });
  });

  describe('configuration', () => {
    beforeEach(function() {
      config = {
        id,
        sources: {
          progressive: [
            {
              mimetype: 'video/mp4',
              url: 'https://www.w3schools.com/tags/movie.mp4'
            }
          ]
        },
        plugins: {
          googleAnalytics: {
            trackingId: 'UA-1234567-89',
            entryId: id,
            entryName,
            uiConfId,
            partnerId
          }
        }
      };
    });

    it('should do nothing for no trackingId given', () => {
      window.dataLayer = null;
      config.plugins.googleAnalytics.trackingId = null;
      setupPlayer(config);
      (!window.dataLayer).should.be.true;
    });

    it('override default category - string', () => {
      config.plugins.googleAnalytics.tracking = {
        category: 'custom string category'
      };
      setupPlayer(config);
      verifyEventParams(dataLayer[dataLayer.length - 1], {category: 'custom string category'});
    });

    it('override default category - function', () => {
      config.plugins.googleAnalytics.tracking = {
        category: function(event = {type: 'custom'}) {
          return `${event.type} ${this.config.uiConfId}`;
        }
      };
      setupPlayer(config);
      verifyEventParams(dataLayer[dataLayer.length - 1], {category: 'custom 123456'});
    });

    it('override default label - string', done => {
      config.plugins.googleAnalytics.tracking = {
        label: 'custom string label'
      };
      setupPlayer(config);
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        verifyEventParams(dataLayer[dataLayer.length - 1], {label: 'custom string label'});
        done();
      });
      player.load();
    });

    it('override default label - function', done => {
      config.plugins.googleAnalytics.tracking = {
        label: function(event) {
          return `${event.type} ${this.config.uiConfId}`;
        }
      };
      setupPlayer(config);
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        verifyEventParams(dataLayer[dataLayer.length - 1], {label: 'medialoaded 123456'});
        done();
      });
      player.load();
    });

    it('override default value - number (round down)', done => {
      config.plugins.googleAnalytics.tracking = {
        value: 10.1
      };
      setupPlayer(config);
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        verifyEventValue(dataLayer[dataLayer.length - 1], 10);
        done();
      });
      player.load();
    });

    it('override default value - function (round up)', done => {
      config.plugins.googleAnalytics.tracking = {
        value: function(event) {
          return event.type.length + this.name.length + 0.9;
        }
      };
      setupPlayer(config);
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        verifyEventValue(dataLayer[dataLayer.length - 1], 27);
        done();
      });
      player.load();
    });

    it('custom event - strings', done => {
      config.plugins.googleAnalytics.tracking = {
        events: {
          MUTE_CHANGE: {
            action: 'mute',
            category: 'custom category',
            label: 'custom label',
            value: 0
          }
        }
      };
      setupPlayer(config);
      player.addEventListener(player.Event.MUTE_CHANGE, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'mute');
        verifyEventParams(dataLayer[dataLayer.length - 1], {category: 'custom category', label: 'custom label'});
        verifyEventValue(dataLayer[dataLayer.length - 1], 0);
        done();
      });
      player.muted = true;
    });

    it('custom event - functions', done => {
      config.plugins.googleAnalytics.tracking = {
        events: {
          MUTE_CHANGE: {
            action: function(event) {
              return `mute ${event.type.length + this.name.length}`;
            },
            category: function(event) {
              return `custom category ${event.type.length + this.name.length}`;
            },
            label: function(event) {
              return `custom label ${event.type.length + this.name.length}`;
            },
            value: function(event) {
              return event.type.length + this.name.length;
            }
          }
        }
      };
      setupPlayer(config);
      player.addEventListener(player.Event.MUTE_CHANGE, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'mute 25');
        verifyEventParams(dataLayer[dataLayer.length - 1], {category: 'custom category 25', label: 'custom label 25'});
        verifyEventValue(dataLayer[dataLayer.length - 1], 25);
        done();
      });
      player.muted = true;
    });

    it('override event', done => {
      config.plugins.googleAnalytics.tracking = {
        events: {
          MEDIA_LOADED: {
            action: 'custom media ready',
            category: 'custom category'
          }
        }
      };
      setupPlayer(config);
      player.addEventListener(player.Event.MEDIA_LOADED, () => {
        verifyEventName(dataLayer[dataLayer.length - 1], 'custom media ready');
        verifyEventParams(dataLayer[dataLayer.length - 1], {category: 'custom category'});
        done();
      });
      player.load();
    });
  });
});


webpack.config.js
'use strict';

const webpack = require('webpack');
const CopyWebpackPlugin = require('copy-webpack-plugin');
const path = require('path');
const PROD = process.env.NODE_ENV === 'production';
const packageData = require('./package.json');

const plugins = [
  new webpack.DefinePlugin({
    __VERSION__: JSON.stringify(packageData.version),
    __NAME__: JSON.stringify(packageData.name)
  })
];

if (PROD) {
  plugins.push(new webpack.optimize.UglifyJsPlugin({sourceMap: true}));
} else {
  plugins.push(
    new CopyWebpackPlugin([
      {
        from: '',
        to: '.'
      }
    ])
  );
}

module.exports = {
  context: __dirname + '/src',
  entry: {
    'pakhshkit-google-analytics': 'index.js'
  },
  output: {
    path: __dirname + '/dist',
    filename: '[name].js',
    library: ['VidiunPlayer', 'plugins', 'googleAnalytics'],
    libraryTarget: 'umd',
    devtoolModuleFilenameTemplate: './google-analytics/[resource-path]'
  },
  devtool: 'source-map',
  plugins: plugins,
  module: {
    rules: [
      {
        test: /\.js$/,
        use: [
          {
            loader: 'babel-loader'
          }
        ],
        exclude: [/node_modules/]
      },
      {
        test: /\.js$/,
        exclude: [/node_modules/],
        enforce: 'pre',
        use: [
          {
            loader: 'eslint-loader',
            options: {
              rules: {
                semi: 0
              }
            }
          }
        ]
      }
    ]
  },
  devServer: {
    contentBase: __dirname + '/src'
  },
  resolve: {
    modules: [path.resolve(__dirname, 'src'), 'node_modules']
  },
  externals: {
    'pakhshkit-js': {
      commonjs: 'pakhshkit-js',
      commonjs2: 'pakhshkit-js',
      amd: 'pakhshkit-js',
      root: ['VidiunPlayer', 'core']
    }
  }
};


src/default-tracking.js
import {Error} from 'pakhshkit-js';

export default {
  category: 'Vidiun Video Events',
  label: function() {
    return `${this.config.partnerId} | ${this.config.uiConfId ? `${this.config.uiConfId} | ` : ''}${this.config.entryId} | '${
      this.config.entryName
    }'`;
  },
  events: {
    MEDIA_LOADED: {
      action: 'media ready'
    },
    FIRST_PLAY: {
      action: 'first play'
    },
    PLAY: {
      action: 'play',
      value: 1
    },
    PAUSE: {
      action: 'pause',
      value: 1
    },
    SEEKED: {
      action: 'seek',
      value: function() {
        return this.player.currentTime;
      }
    },
    ENDED: {
      action: 'ended'
    },
    CHANGE_SOURCE_ENDED: {
      action: 'change media',
      value: 1,
      condition: function() {
        return !this._firstEntry;
      }
    },
    ENTER_FULLSCREEN: {
      action: 'enter full screen',
      value: 1
    },
    EXIT_FULLSCREEN: {
      action: 'exit full screen',
      value: 1
    },
    ERROR: {
      action: 'error',
      category: 'Vidiun Video Error',
      label: function(error) {
        // eslint-disable-next-line no-unused-vars
        return Object.entries(Error.Code).find(([name, code]) => {
          return code === error.payload.code;
        })[0];
      },
      condition: function(error) {
        return error.payload.severity === Error.Severity.CRITICAL;
      }
    }
  }
};


src/google-analytics.js
// @flow
import {BasePlugin, Utils} from 'pakhshkit-js';
import defaultTracking from './default-tracking';

const WIDGET_LOADED_ACTION: string = 'widget loaded';
const PCT_25_ACTION: string = '25% watched';
const PCT_50_ACTION: string = '50% watched';
const PCT_75_ACTION: string = '75% watched';
const PCT_100_ACTION: string = '100% watched';

/**
 * Your class description.
 * @classdesc
 */
export default class GoogleAnalytics extends BasePlugin {
  /**
   * The default configuration of the plugin.
   * @type {Object}
   * @static
   */
  static defaultConfig: Object = {
    tracking: defaultTracking
  };

  /**
   * @static
   * @public
   * @returns {boolean} - Whether the plugin is valid.
   */
  static isValid(): boolean {
    return true;
  }

  /**
   * The gtag lib url.
   * @type {string}
   * @static
   */
  static GTAG_LIB_URL: string = '//www.googletagmanager.com/gtag/js';

  /**
   * Indicate whether time percent event already sent
   * @private
   */
  _timePercentEvent: {[event: string]: boolean} = {
    PLAY_REACHED_25: false,
    PLAY_REACHED_50: false,
    PLAY_REACHED_75: false,
    PLAY_REACHED_100: false
  };

  /**
   * Indicate whether its the first entry that is played
   * @private
   */
  _firstEntry: boolean = true;

  /**
   * @constructor
   * @param {string} name - The plugin name.
   * @param {Player} player - The player instance.
   * @param {Object} config - The plugin config.
   */
  constructor(name: string, player: Player, config: Object) {
    super(name, player, config);
    if (this.config.trackingId) {
      this._init();
      this._addBindings();
      this._sendEvent({
        action: WIDGET_LOADED_ACTION,
        category: this._getValue(this.config.tracking.category)
      });
    } else {
      this.logger.warn('No Google Analytics tracking ID provided. Tracking aborted');
    }
  }

  /**
   * _gtag - update the dataLayer for gtag manager
   * @private
   * @returns {void}
   */
  _gtag() {
    window.dataLayer.push(arguments);
  }

  /**
   * _init
   * @private
   * @returns {void}
   */
  _init(): void {
    if (!window.google_tag_manager) {
      Utils.Dom.loadScriptAsync(`${GoogleAnalytics.GTAG_LIB_URL}?id=${this.config.trackingId}`).then(() => {
        this.logger.debug('Google gtag library has loaded successfully');
      });
    }
    window.dataLayer = window.dataLayer || [];
    // $FlowFixMe
    this._gtag('js', new Date());
    // $FlowFixMe
    this._gtag('config', this.config.trackingId);
  }

  /**
   * _getValue - returns the value itself or the returned value if it's a function
   * @param {string | number | function} val - the value to return or calculate
   * @param {?Object} event - passing to the function if exist
   * @returns {?string | ?number} - the returned value
   * @private
   */
  _getValue(val: any, event: ?Object): any {
    try {
      return typeof val === 'function' ? val.call(this, event) : val;
    } catch (e) {
      this.logger.error(e);
      return null;
    }
  }

  /**
   * _addBindings
   * @private
   * @returns {void}
   */
  _addBindings(): void {
    Object.entries(this.config.tracking.events).forEach(([eventName, eventParams]) => {
      this.eventManager.listen(this.player, this.player.Event[eventName], event => {
        const shouldSendEvent = eventParams => {
          return eventParams.action && (typeof eventParams.condition === 'function' ? eventParams.condition.call(this, event) : true);
        };
        if (eventParams && typeof eventParams === 'object' && shouldSendEvent(eventParams)) {
          const customCategory = this._getValue(eventParams.category, event);
          const customLabel = this._getValue(eventParams.label, event);
          const customValue = this._getValue(eventParams.value, event);
          const eventObj = {
            action: this._getValue(eventParams.action, event),
            category: typeof customCategory === 'string' ? customCategory : this._getValue(this.config.tracking.category, event),
            label: typeof customLabel === 'string' ? customLabel : this._getValue(this.config.tracking.label, event),
            value: typeof customValue === 'number' ? customValue : this._getValue(this.config.tracking.value, event)
          };
          this._sendEvent(eventObj);
        }
      });
    });
    this.eventManager.listen(this.player, this.player.Event.TIME_UPDATE, this._sendTimePercentAnalytic.bind(this));
    this.eventManager.listenOnce(this.player, this.player.Event.CHANGE_SOURCE_ENDED, () => (this._firstEntry = false));
  }

  /**
   * Send time percent analytic
   * @param {Object} event - 'timeupdate' event
   * @private
   * @return {void}
   */
  _sendTimePercentAnalytic(event: Object): void {
    const getPctEventParams = () => {
      return {
        category: this._getValue(this.config.tracking.category, event),
        label: this._getValue(this.config.tracking.label, event),
        value: this.player.currentTime
      };
    };
    if (this.player.config.sources.type !== this.player.MediaType.LIVE) {
      const percent = this.player.currentTime / this.player.duration;
      if (!this._timePercentEvent.PLAY_REACHED_25 && percent >= 0.25) {
        this._timePercentEvent.PLAY_REACHED_25 = true;
        this._sendEvent({
          action: PCT_25_ACTION,
          ...getPctEventParams()
        });
      }
      if (!this._timePercentEvent.PLAY_REACHED_50 && percent >= 0.5) {
        this._timePercentEvent.PLAY_REACHED_50 = true;
        this._sendEvent({
          action: PCT_50_ACTION,
          ...getPctEventParams()
        });
      }
      if (!this._timePercentEvent.PLAY_REACHED_75 && percent >= 0.75) {
        this._timePercentEvent.PLAY_REACHED_75 = true;
        this._sendEvent({
          action: PCT_75_ACTION,
          ...getPctEventParams()
        });
      }
      if (!this._timePercentEvent.PLAY_REACHED_100 && percent >= 1) {
        this._timePercentEvent.PLAY_REACHED_100 = true;
        this._sendEvent({
          action: PCT_100_ACTION,
          ...getPctEventParams()
        });
      }
    }
  }

  /**
   * _sendEvent
   * @param {Object} event - event to send
   * @private
   * @returns {void}
   */
  _sendEvent(event: Object) {
    if (event.action) {
      const eventProps = {};
      eventProps['event_category'] = event.category;
      if (event.label) {
        eventProps['event_label'] = event.label;
      }
      if (typeof event.value === 'number') {
        eventProps['value'] = Math.round(event.value);
      }
      this.logger.debug(`${event.action} event sent`, eventProps);
      // $FlowFixMe
      this._gtag('event', event.action, eventProps);
    }
  }

  /**
   * Destroys the plugin.
   * @override
   * @public
   * @returns {void}
   */
  destroy(): void {
    this.eventManager.destroy();
  }

  /**
   * Resets the plugin.
   * @override
   * @public
   * @returns {void}
   */
  reset(): void {
    this._timePercentEvent = {
      PLAY_REACHED_25: false,
      PLAY_REACHED_50: false,
      PLAY_REACHED_75: false,
      PLAY_REACHED_100: false
    };
  }
}
